Rename Stage model to Lesson

uponthesun opened the pull request:
This is (close to) the minimum set of changes needed to rename our `Stage` model to `Lesson`. This is purely an application code change; none of the database table or column names will change. We want to do this to align the terms in our codebase better with the terms used by the curriculum team. Among all the terms which differ, we're prioritizing "Stage" since anecdotally, it's the most confusing.

This change mostly consists of:
* Renaming `stage.rb` to `lesson.rb`
* Updating associations to refer to `:lesson`/`:lessons` instead of `:stage`/`:stages`, explicitly specifying the foreign key name when necessary.
* Updating references to the `.stage`/`.stages` fields of model objects to `.lesson`/`.lessons`

This change should not change:
* i18n data or any interactions with it
* the Script DSL format
* `Script.summarize`, which returns data used by some of the views
* any other fields of any models that have "stage" in the name
* most function and local variable names (some got changed since I was initially trying to get everything in one go. That quickly proved impractical, but I didn't revert everything I'd already changed.)

### Deployment strategy
I think this should work without any special care need for the deployment, since the database / seed data / i18n data is staying the same. Please help think of anything I could be missing though!

### Future work
I'm not sure if / when we'll want to change the other parts mentioned above, but I think each can be done independently.

## Links

<!--
  Any relevant links to external resources; ie, specification documents, jira
  items, related PRs, honeybadger errors, etc
-->

https://docs.google.com/document/d/1bJPhdwX7QZHRT12dQBJcXrYDKGSdobMp1c_6qAuvUEM/edit#bookmark=id.fc54sb9b79su

## Testing story

* Unit and UI tests pass on Drone
* Seeding works (this is tested by Drone UI tests, which have to seed)

Interested in thoughts on whether we need to specifically testing Script DSL and i18n beyond the CI tests, and if so how.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

dmcavoy commented on dashboard/test/controllers/scripts_controller_test.rb:
```diff
@@ -663,7 +663,7 @@ class ScriptsControllerTest < ActionController::TestCase
     get :show, params: {id: 'test-fixture-visible-after'}
     assert_response :success
-    assert response.body.include? 'The lesson Stage 1 will be visible after'
+    assert response.body.include? 'The lesson stage 1 will be visible after'
```
Should this be updated to Lesson or not changed?

uponthesun replied:
Yea I was a little careless with these, haha. I can fix these to Lesson

uponthesun replied:
On second thought, I'd like to leave these as-is until we follow up with the work to change the script DSL format. It's not ideal that I renamed them uselessly but I think we'll be changing them again soon so I'd rather not change it again.

dmcavoy replied:
Ok thats fine

dmcavoy commented on dashboard/test/factories/factories.rb:
```diff
@@ -730,18 +730,18 @@
 
-  factory :stage do
-    sequence(:name) {|n| "Bogus Stage #{n}"}
+  factory :lesson, class: 'Lesson' do
```
Does the class need to be specified (more of a question for my own understanding)?

uponthesun replied:
Oh, good catch, doesn't look like it's needed. Will remove.

dmcavoy commented on dashboard/test/fixtures/duplicate_scripts/test-fixture-3-stages.script:
```diff
@@ -1,7 +1,7 @@
-stage 'Stage3'
+stage 'stage3'
```
Should this be updated to lesson?

dmcavoy commented on dashboard/test/fixtures/duplicate_scripts/test-fixture-3-stages.script:
```diff
@@ -1,7 +1,7 @@
 level 'Level 5'
 
-stage 'Stage1'
+stage 'stage1'
```
Should this be updated to lesson?

dmcavoy commented on dashboard/test/fixtures/duplicate_scripts/test-fixture.script:
```diff
@@ -1,3 +1,3 @@
-stage 'Stage1'
+stage 'stage1'
```
Should this be updated to lesson?

dmcavoy commented on dashboard/test/fixtures/login_required.script:
```diff
@@ -1,4 +1,4 @@
 login_required 'true'
 
-stage 'Stage1'
+stage 'stage1'
```
lesson1?

dmcavoy commented on dashboard/test/fixtures/test-fixture-3-stages.script:
```diff
@@ -1,10 +1,10 @@
-stage 'Stage1'
+stage 'stage1'
```
lesson1? (Also other stages in this script file)

dmcavoy commented on dashboard/test/fixtures/test-fixture-experiments.script:
```diff
@@ -1,4 +1,4 @@
-stage 'Stage1'
+stage 'stage1'
```
lesson1?

dmcavoy commented on dashboard/test/fixtures/test-fixture-variants.script:
```diff
@@ -1,4 +1,4 @@
-stage 'Stage1'
+stage 'stage1'
```
lesson1?

dmcavoy commented on dashboard/test/fixtures/test-fixture-versioned-1801.script:
```diff
@@ -2,7 +2,7 @@ family_name 'versioned'
 is_stable true
 
-stage 'Stage1'
+stage 'stage1'
```
lesson1?

dmcavoy commented on dashboard/test/fixtures/test-fixture-visible-after.script:
```diff
@@ -1,2 +1,2 @@
-stage 'Stage 1', visible_after: '2020-04-01 08:00:00 -0700'
+stage 'stage 1', visible_after: '2020-04-01 08:00:00 -0700'
```
lesson 1?

dmcavoy commented on dashboard/test/fixtures/test-fixture.script:
```diff
@@ -1,10 +1,10 @@
 script_announcements [{"notice"=>"foo", "details"=>"bar", "link"=>"", "type"=>"information"}]
 
-stage 'Stage1'
+stage 'stage1'
```
lesson1? also other stage lesson2?

dmcavoy commented on dashboard/test/fixtures/test-unplugged.script:
```diff
@@ -1,4 +1,4 @@
-stage 'Stage1'
+stage 'stage1'
```
lesson1?

dmcavoy commented on dashboard/test/models/script_test.rb:
```diff
@@ -67,7 +67,7 @@ def populate_cache_and_disconnect_db
     script = scripts[0]
     assert_equal 'Level 1', script.levels[0].name
-    assert_equal 'Stage2', script.script_levels[3].stage.name
+    assert_equal 'stage2', script.script_levels[3].lesson.name
```
Update to lesson?

dmcavoy commented on dashboard/test/models/script_test.rb:
```diff
@@ -154,44 +154,44 @@ def populate_cache_and_disconnect_db
+    first = scripts[0].lessons[0]
+    second = scripts[0].lessons[1]
+    third = scripts[0].lessons[2]
+    assert_equal 'stage1', first.name
```
Update to lesson1 (and following stages too)?

dmcavoy commented on dashboard/test/models/script_test.rb:
```diff
@@ -154,44 +154,44 @@ def populate_cache_and_disconnect_db
     assert_equal 2, second.absolute_position
-    assert_equal 'Stage3', first.name
-    assert_equal 'Stage1', second.name
+    assert_equal 'stage3', first.name
```
Update to lesson?

dmcavoy commented on dashboard/test/models/script_test.rb:
```diff
@@ -681,29 +681,29 @@ def populate_cache_and_disconnect_db
 
     assert_equal 1, summary[:stages].count
-    assert_nil summary[:peerReviewStage]
+    assert_nil summary[:peerReviewlesson]
```
peerReviewLesson (upper case L)?

dmcavoy replied:
Also was this updated to this name because searching for it did not find anything and I see another one down below that is still peerReviewStage

davidsbailey replied:
apologies if I am missing something, but this change in particular looks bogus because the summary may still contain a peerReviewStage: https://github.com/code-dot-org/code-dot-org/blob/481131a6b2ec0f1eacbf6853271e2c7d45cfc3d2/dashboard/app/models/script.rb#L1309  and the test doesn't catch it because all undefined properties on hashes are `nil`

uponthesun replied:
Good catch, that probably would've broken the test's ability to detect problems... I'll change it back to peerReviewStage.

dmcavoy commented on dashboard/test/models/script_test.rb:
```diff
@@ -719,7 +719,7 @@ def populate_cache_and_disconnect_db
 
     assert_equal 2, summary[:stages].count
-    assert_equal expected_peer_review_stage, summary[:peerReviewStage]
+    assert_equal expected_peer_review_lesson, summary[:peerReviewStage]
```
Other spot where peerReviewStage shows up - see comment on peerReviewlesson

uponthesun replied:
Yea, this is the correct key, I'll fix the other one.

dmcavoy commented on dashboard/test/models/script_test.rb:
```diff
@@ -1496,7 +1495,7 @@ class SummarizeVisibleAfterScriptTests < ActiveSupport::TestCase
     # variants / active / endvariants structure is correct.
     new_dsl_regex = <<-SCRIPT
-stage 'Stage1'
+stage 'stage1'
```
lesson1?

dmcavoy commented on pegasus/src/curriculum_course.rb:
```diff
@@ -1,8 +1,7 @@
-
-Lesson = Struct.new(:number, :url, :title, :description, :path)
-
 class CurriculumCourse
```
What does CurriculumCourse do?

uponthesun replied:
I don't have context here but it looks like it provides some metadata about courses to various places in Pegasus. I had to change this because the Lesson defined here were conflicting with the model definition.

davidsbailey commented on dashboard/app/models/script.rb:
```diff
@@ -887,26 +887,26 @@ def self.setup(custom_files, new_suffix: nil, new_properties: {})
-  def self.add_script(options, raw_stages, new_suffix: nil, editor_experiment: nil)
-    raw_script_levels = raw_stages.map {|stage| stage[:scriptlevels]}.flatten
+  def self.add_script(options, raw_lessons, new_suffix: nil, editor_experiment: nil)
+    raw_script_levels = raw_lessons.map {|stage| stage[:scriptlevels]}.flatten
     script = fetch_script(options)
     script.update!(hidden: true) if new_suffix
     chapter = 0
     stage_position = 0; script_level_position = Hash.new(0)
-    script_stages = []
+    script_lessons = []
     script_levels_by_stage = {}
```
I have mixed feelings here because these particular renames leave us in a bit of limbo. would you be willing to at least also rename `script_levels_by_stage` and `raw_stage`, since those look like fairly clean renames, and that would eliminate some of the back and forth between lesson and stage?

uponthesun replied:
Ok, I'll go through this method and rename things one by one so everything is correct and consistent. I think we can rename everything except for things used in keys.

davidsbailey commented on dashboard/app/models/lesson.rb:
```diff
@@ -23,17 +23,19 @@
-  has_many :script_levels, -> {order('position ASC')}, inverse_of: :stage
-  has_one :plc_learning_module, class_name: 'Plc::LearningModule', inverse_of: :stage, dependent: :destroy
-  belongs_to :script, inverse_of: :stages
+  has_many :script_levels, -> {order('position ASC')}, inverse_of: :lesson, foreign_key: 'stage_id'
+  has_one :plc_learning_module, class_name: 'Plc::LearningModule', inverse_of: :lesson, foreign_key: 'stage_id', dependent: :destroy
```
I love this approach to renaming the associations without having to rename the DB table columns. Nice choice! We can wait and see if it will become worth renaming the tables... but I am guessing that it is unlikely.

davidsbailey commented on pegasus/test/test_curriculum_course.rb:
```diff
@@ -32,7 +32,7 @@ def setup
     it 'should get all lessons' do
       assert_equal(18, @course.get_lessons.count)
-      assert(@course.get_lessons.first.is_a?(Lesson))
+      assert(@course.get_lessons.first.is_a?(CurriculumCourse::Lesson))
```
I am just curious, did these change because of a test failure due to a conflict between the new Lesson class in Rails? I am surprised to see that pegasus tests might have access to rails classes. or did rubymine take care of this as part of an automated rename? either way, extra specificity seems fine.

uponthesun replied:
This was actually causing issues in dashboard because somehow the types defined in the `curriculum_course` file in the top-level namespace were overriding the ones defined in the model file sometimes. I don't remember the exact details, but moving the pegasus definitions of Lesson and Unit into the `CurriculumCourse` class fixed it.

uponthesun commented:
@dmcavoy @davidsbailey I think I've addressed the comments. I also tested locally that you can add and remove lessons from a script in levelbuilder mode.

dmcavoy commented on dashboard/app/models/script.rb:
```diff
@@ -919,8 +919,8 @@ def self.add_script(options, raw_stages, new_suffix: nil, editor_experiment: nil
-      stage_flex_category = nil
-      stage_lockable = nil
+      flex_category = nil
+      lockable = nil
```
I think it might be good to call this lesson_lockable because there are variables floating around here for lesson and levels and it will just make it clearer

dmcavoy replied:
I don't think its a huge deal tho so if you disagree thats totally fine.

uponthesun replied:
I didn't care a lot about this since neither seems clearly better to me but... I'll leave it as if that's okay with you 😅
